File: frontend/src/api/articles.ts
import { get } from './client'
import type { Article } from '../types/article'

interface ArticlesResponse {
  articles: Article[]
  offset: number
  limit: number
}

export function fetchArticles(params: {
  date?: string
  source?: string
  limit?: number
  offset?: number
}): Promise<ArticlesResponse> {
  const query = new URLSearchParams()
  if (params.date) query.set('date', params.date)
  if (params.source) query.set('source', params.source)
  if (params.limit) query.set('limit', String(params.limit))
  if (params.offset) query.set('offset', String(params.offset))
  return get<ArticlesResponse>(`/articles?${query}`)
}

File: backend/app/api/v1/articles.py
from fastapi import APIRouter, Query
from app.database import supabase

router = APIRouter(prefix="/articles", tags=["articles"])


@router.get("")
def list_articles(
    date: str = Query(None),
    source: str = Query(None),
    limit: int = Query(20, le=50),
    offset: int = Query(0),
):
    try:
        query = supabase.table("articles").select("*").order("published_at", desc=True)

        if date:
            query = query.eq("published_at::date", date)
        if source:
            query = query.eq("source", source)

        result = query.range(offset, offset + limit - 1).execute()
        return {"articles": result.data or [], "offset": offset, "limit": limit}
    except Exception as e:
        return {"error": str(e)}
